src/lua/protobuf.lua

```diff
@@ -0,0 +1,608 @@
+    -- but not defined in protocol. Allows a type be defined after usage
+    for def_type, declared in pairs(declarations) do
+        if not declared then
+            error(('Type %q is not declared'):format(def_type))
+        end
+    end
+    return setmetatable(res, protocol_mt)
+end
+
+local function is_number64(value)
+    if type(value) == 'cdata' and (ffi.istype(int64_t, value) or
+        ffi.istype(uint64_t, value)) then
+        return true
+    end
+end
+
+local function validate_length(value)
+    local MAX_LEN = 2^32
+    if string.len(value) > MAX_LEN then
+        error("Too long string to be encoded")
+    end
+end
+
+local function encode_len(value, field_id)
+    local wire_type = 2
+    return string.format('%s%s%s',
+        wireformat.encode_uint(bit.lshift(field_id, 3) + wire_type),
+        wireformat.encode_uint(string.len(value)),
+        value)
+end
+
+local encode_field
+
+local encode
+
+local function encode_repeated(protocol, field_def, data)
+    local buf = {}
+    local scalar_def = scalars[field_def.type]
+    if type(data) ~= 'table' then
+        error('For repeated fields table data are needed')
+    elseif scalar_def ~= nil and scalar_def.encode_as_packed then
+        for _, value in pairs(data) do
+            table.insert(buf, string.sub(encode_field(protocol, field_def,
+                value, true), 2))
+        end
+        return encode_len(table.concat(buf), field_def.id)
+    else
+        for _, value in pairs(data) do
+            table.insert(buf, encode_field(protocol, field_def, value, true))
+        end
+        return table.concat(buf)
+    end
+end
+
+local function check_integer(field_def, value)
+    if (type(value) == 'number' and math.ceil(value) ~= value) then
+        error(('Input number value %f for %q is not integer'):format(
+            value, field_def.name))
+    elseif type(value) == 'cdata' and not is_number64(value) then
+        error(('Input cdata value %q for %q field is not integer'):format(
+            ffi.typeof(value), field_def.name))
+    end
+end
+
+local function encode_enum(value, list_of_values, field_id, field_type)
+    if type(value) ~= 'number' and list_of_values[value] == nil then
+        error(('%q is not defined in %q enum'):format(value, field_type))
+    end
+    -- According to open enums semantics unknown enum values are encoded as
+    -- numeric identifier. https://protobuf.dev/programming-guides/enum/
+    if type(value) == 'number' then
+        local field_def = {type = 'int32', id = field_id}
+        scalars['int32'].validate(value, field_def)
+        return scalars['int32'].encode(value, field_id)
+    else
+        return scalars['int32'].encode(list_of_values[value], field_id)
+    end
+end
+
+encode_field = function(protocol, field_def, value, ignore_repeated)
+    local scalar_def = scalars[field_def.type]
+    local non_scalars_def = protocol[field_def.type]
+    if field_def.repeated and ignore_repeated == false then
```

Comment:
Style nit: Don't check booleans for `== true` or `== false`. Use them directly in an if-condition or a logical expression.

Reply:
Fixed, changed it with `and not ignore_repeated`